llvm/lib/Target/PowerPC/PPCPreEmitPeephole.cpp

```diff
@@ -109,6 +109,93 @@ static bool hasPCRelativeForm(MachineInstr &Use) {
           MachineFunctionProperties::Property::NoVRegs);
     }
 
+    // The funtion will simply the zeroing accumulator and spilling instrcutions
+    // into simple xxlxor and spilling instrcuctions.
+    // From:
+    // setaccz acci
+    // xxmfacc acci
+    // stxv vsr(i*4+0), D(1)
+    // stxv vsr(i*4+1), D-16(1)
+    // stxv vsr(i*4+2), D-32(1)
+    // stxv vsr(i*4+3), D-48(1)
+
+    // To:
+    // xxlxor vsr(i*4), 0, 0
+    // stxv vsr(i*4), D(1)
+    // stxv vsr(i*4), D-16(1)
+    // stxv vsr(i*4), D-32(1)
+    // stxv vsr(i*4), D-48(1)
+    bool
+    OptimizeZeroingAccumulatorSpilling(MachineBasicBlock &MBB,
+                                       const TargetRegisterInfo *TRI) const {
+      bool changed = false;
+      for (auto BBI = MBB.instr_begin(); BBI != MBB.instr_end(); ++BBI) {
+        if (BBI->getOpcode() != PPC::XXSETACCZ)
+          continue;
+
+        Register ACCZReg = BBI->getOperand(0).getReg();
+
+        DenseSet<MachineInstr *> InstrsToErase;
+        InstrsToErase.insert(&*BBI++);
+
```

Comment:
I agree. 
We want this to be a general solution. I'm not sure that the approach of putting this so late in the pre-emit peephole is a good idea. When are these stores generated? Is this a spill? Is it just a store of a zeroed ACC register?

It may be easier to fix this the first time when the stores are generated because then the instruction scheduling won't be a problem and we just generate the instructions we want.

Reply:
the patch is for issue https://github.ibm.com/compiler/wyvern/issues/16624, it is a spill, there is another solution to  whether there is XXSETACCZ instruction in the function https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/PowerPC/PPCRegisterInfo.cpp#L1313 

if there is the XXSETACCZ  instruction , we do not do 
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/PowerPC/PPCRegisterInfo.cpp#L1336 
`BuildMI(MBB, II, DL, TII.get(PPC::XXMFACC), SrcReg).addReg(SrcReg);`

and in the function  [static void spillRegPairs](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/PowerPC/PPCRegisterInfo.cpp#L1238) 
we change the code

```
addFrameReference(BuildMI(MBB, II, DL, TII.get(PPC::STXV))
                        .addReg(Reg + i, getKillRegState(IsKilled)),
                    FrameIndex, Offset);
```
                    
  to 
  
  ```
addFrameReference(BuildMI(MBB, II, DL, TII.get(PPC::STXV))
                        .addReg(Reg , getKillRegState(IsKilled)),
                    FrameIndex, Offset);
```

but I do not think it is good place to put optimize in the function `PPCRegisterInfo::lowerACCSpilling`

Reply:
> it is not necessary there must be 4 stores...
the patch is accumulator spill, it must has 4 stores.

> limiting the case that XXSETACCZ and XXMFACC must be adjacent is not a good idea...
OK, I will deal the XXSETACCZ and XXMFACC are not adjacent.

> limitting the case that the four stores must be followed XXMFACC is also not good.
the patch is accumulator spill, in the function  [PPCRegisterInfo::lowerACCSpilling](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/PowerPC/PPCRegisterInfo.cpp#L1313). the XXMFACC is followed by 4 store, but I can deal with XXMFACC is not followed by 4 store any ways.

> The loop structure also seems not able to handle case
I will deal with these case.